pandas/core/strings.py

```diff
@@ -418,38 +418,123 @@ def _get_single_group_name(rx):
         return None
 
 
-def str_extract(arr, pat, flags=0):
+def _groups_or_na_fun(regex):
+    """Used in both extract_noexpand and extract_frame"""
+    if regex.groups == 0:
+        raise ValueError("pattern contains no capture groups")
+    empty_row = [np.nan] * regex.groups
+    def f(x):
+        if not isinstance(x, compat.string_types):
+            return empty_row
+        m = regex.search(x)
+        if m:
+            return [np.nan if item is None else item for item in m.groups()]
+        else:
+            return empty_row
+    return f
+
+
+def _str_extract_noexpand(arr, pat, flags=0):
     """
     Find groups in each string in the Series using passed regular
-    expression.
+    expression. This function is called from
+    str_extract(expand=False), and can return Series, DataFrame, or
+    Index.
+
+    """
+    from pandas.core.frame import DataFrame
+    from pandas.core.index import Index
+
+    regex = re.compile(pat, flags=flags)
+    groups_or_na = _groups_or_na_fun(regex)
```

Comment:
To avoid repetition, `_groups_or_na_fun(regex)` is used in both `_str_extract_noexpand` and `_str_extract_frame`